New sample - Snap geometry edits with utility network rules

shellygill opened the pull request:
## Description

PR to add a new Kotlin sample _"Snap geometry edits with utility network rules"_ in `Utility Networks` category.

## Links and Data

- Feature team issue https://devtopia.esri.com/runtime/bistromath/issues/3070
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/) Job for this PR has been run
  - [ ] link:  https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/96 ~/95 /92 /91 /90~

## What To Review

-  Putting this up for a early review stage as discussed with @TADraeseke - some specific discussion points noted below

## How to Test

Run the sample on the sample viewer or the repo.

## To Discuss

- We talked about how best to expose the snap source settings state list to Compose - here I rebuild the list every time there is an `SnapSourceSettings.enabled` change - there's only 3 items so I don't expect this to be a performance problem, but happy to take suggestions on the better approach here.
- I have two separate fields `symbols` and `symbolSwatches` mutable maps - is there a better way to do this (use a Pair?), and ideally one that avoids the non-null assertion I have later on (lazy initialiser?) - one of the maps can be immediately initialised, but the other must be done later on during app usage.
- `createSwatch `- I found the line symbol rather thin and difficult to see using a more expected DPs size, so I played around a little with the values of `width`/`height` and `screenScale` here - any advice to improve this?

<!-- OPTIONAL
## Screenshots
-->

shellygill commented on gradle/libs.versions.toml:
```diff
@@ -1,7 +1,7 @@
 
 # ArcGIS Maps SDK for Kotlin version
-arcgisMapsKotlinVersion = "200.7.0-4516"
+arcgisMapsKotlinVersion = "200.7.0-4541"
```
I needed to update the existing v.Next version in order to pick up final API changes.

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+        val pipeLayer = SubtypeFeatureLayer(
+            geodatabase.getFeatureTable(pipelineLayerName) as ArcGISFeatureTable
+        )
+        pipeLayer.load().getOrThrow()
```
```suggestion
        pipeLayer.load().getOrElse {
            messageDialogVM.showMessageDialog(
                "Error loading pipeline layer",
                it.message.toString()
            )
        }
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+                application.getString(R.string.device_layer_name)
+            ) as ArcGISFeatureTable
+        )
+        deviceLayer.load().getOrThrow()
```
I'd use getOrElse here too

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,13 @@
+    <string name="snap_geometry_edits_with_utility_network_rules_app_name">Snap geometry edits with utility network rules</string>
+    <string name="naperville_geodatabase">/NapervilleGasUtilities.geodatabase</string>
+    <string name="pipeline_layer_name">PipelineLine</string>
+    <string name="service_pipe_layer_name">Service Pipe</string>
+    <string name="distribution_pipe_name">Distribution Pipe</string>
+    <string name="device_layer_name">PipelineDevice</string>
+    <string name="excess_flow_valve_name">Excess Flow Valve</string>
+    <string name="junction_layer_name">PipelineJunction</string>
+    <string name="controllable_tee_name">Controllable Tee</string>
```
Ok let's not worry then. Broadly doc is supposed to be sentence case, I believe. So we just try to carry that into samples too. Not a big deal, especially if it matches the data

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/screens/SnapGeometryEditsWithUtilityNetworkRulesScreen.kt:
```diff
@@ -0,0 +1,129 @@
+                    mapViewProxy = mapViewModel.mapViewProxy,
+                    onSingleTapConfirmed = mapViewModel::identify
+                )
+                SnapSourcesPanel(
```
The modal bottom sheet won't change the map size since it opens up over top of the map. We use that pattern in many compose samples. And I agree that changes in map view size are really annoying.

Where to put the instructions...hmm...maybe they fill the current whitespace below the map view (and shrink that space to just the one line of text) and the modal sheet goes over top? Is the only instruction "Tap a point feature to edit"? If so, I think it's ok just to have that instruction in the README

shellygill replied:
Ah, got you, I will make follow-up changes then for this.

> Is the only instruction "Tap a point feature to edit"? 

Yes - I think because only point features will work, not the lines, the design was to specifically highlight this in the sample UI, so I'll try the approach you suggested.

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/screens/SnapGeometryEditsWithUtilityNetworkRulesScreen.kt:
```diff
@@ -0,0 +1,129 @@
+                Row(modifier = Modifier.padding(horizontal = 16.dp)) {
+                    Text(text = "AssetType: ")
+                    Text(text = mapViewModel.assetTypeNameState.collectAsState().value)
+                }
```
Could it go in a box that goes over top of the map to avoid changes in the map view size? Or could use the composable callout maybe? https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/blob/v.next/samples/manage-features/src/main/java/com/esri/arcgismaps/sample/managefeatures/screens/ManageFeaturesScreen.kt#L87

@shubham7109 often has good UI ideas around things like this

shellygill replied:
This has been resolved by Shubhams changes.

TADraeseke commented:
`We talked about how best to expose the snap source settings state list to Compose - here I rebuild the list every time there is an SnapSourceSettings.enabled change - there's only 3 items so I don't expect this to be a performance problem, but happy to take suggestions on the better approach here.`

I think it's fine--as you say it's a list of 3. The Kotlin recommended approach is to remove and insert at the specified index which is mutating. In fact data classes come with a .copy function for this reason where you can pass the property that changes as a parameter to the copy function. So in your code it might look something like `_snapSourcePropertyList[index] = _snapSourcePropertyList[index].copy(snapSourceSettings = newSourceSetting)` NOTE: I'm just hacking this line in the text box of github, so syntax may not be precise.


`I have two separate fields symbols and symbolSwatches mutable maps - is there a better way to do this (use a Pair?), and ideally one that avoids the non-null assertion I have later on (lazy initialiser?) - one of the maps can be immediately initialised, but the other must be done later on during app usage.`

I agree repeated null checks are annoying, the only Kotlin alternative is to use a `lateinit`, but then you should really be checking .isInitialized repeatedly, which defeats the whole point. I think we just lean into the null checks here as Kotlin keeping us right.

Could do a list of pairs as you say, but I think a Map seems an entirely appropriate data structure here.

`createSwatch - I found the line symbol rather thin and difficult to see using a more expected DPs size, so I played around a little with the values of width/height and screenScale here - any advice to improve this?`

I had to do the same with [another sample](https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/blob/v.next/samples/create-symbol-styles-from-web-styles/src/main/java/com/esri/arcgismaps/sample/createsymbolstylesfromwebstyles/components/CreateSymbolStylesFromWebStylesViewModel.kt#L132). I think what you're doing looks reasonable to me

shubham7109 reviewed (commented):
@shellygill fascinating sample design! 

I took a look mainly into the Compose UI aspect for this sample, and worked on some enhancements to help make the workflow simplified following Kotlin & Compose patterns. 

Check out the changes: https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/335

I definitely took some liberty with the UI design, so feel free to tweak them as it feels helpful.

shellygill commented:
> I took a look mainly into the Compose UI aspect for this sample, and worked on some enhancements to help make the workflow simplified following Kotlin & Compose patterns.
> Check out the changes: https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/335
> I definitely took some liberty with the UI design, so feel free to tweak them as it feels helpful.

@shubham7109 thanks so much for your improvements - I think they look good to me, I think the workflow tweaks work well, and I dont think it strays too much from the original sample intent myself, it worked well for me when trying it out (and the state manangement is obviously more in line with how Kotlin samples should be working (I have more to learn here). I'm going to merge these straight in, and then update this branch as it's become out of date with v.next.

shellygill commented:
Apparently I also forgot to hit the "ready for review" button, fixing that now too.

shellygill commented:
Thanks @TADraeseke and @shubham7109 for your discussion and improvements for the UI. I've pulled in all those changes, updated against v.next, and fixed errors being reported in the metadata & readme (after a few tries 😆 ).

TADraeseke commented:
I'll do first review on this tomorrow

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/README.md:
```diff
@@ -0,0 +1,60 @@
+4. Create a `GeometryEditor`, set `GeometryEditor.tool` to a `ReticleVertexTool`, and connect the editor to the map view.
+5. When editing a feature:
+
+    a. Call `SnapRules.CreateAsync(UtilityNetwork, UtilityAssetType)` to get the snap rules associated with a given `UtilityAssetType`.
```
```suggestion
    a. Call `SnapRules.create(UtilityNetwork, UtilityAssetType)` to get the snap rules associated with a given `UtilityAssetType`.
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/README.md:
```diff
@@ -0,0 +1,60 @@
+
+    a. Call `SnapRules.CreateAsync(UtilityNetwork, UtilityAssetType)` to get the snap rules associated with a given `UtilityAssetType`.
+
+    b. Use `SyncSourceSettings(SnapRules, SnapSourceEnablingBehavior.setFromRules)` to populate the `SnapSettings.SourceSettings` with `SnapSourceSettings` enabling the sources with rules.
```
```suggestion
    b. Use `syncSourceSettings(SnapRules, SnapSourceEnablingBehavior.setFromRules)` to populate the `snapSettings.sourceSettings` with `SnapSourceSettings` enabling the sources with rules.
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+            }
+
+            // Hide the selected feature
+            (feature.featureTable!!.layer as? FeatureLayer)?.setFeatureVisible(feature, false)
```
```suggestion
            (feature.featureTable?.layer as? FeatureLayer)?.setFeatureVisible(feature, false)
```

Can just be a null check here

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+    }
+
+    /**
+     * Creates SnapRules based on the given asset type, synchronizes the snap sources using these
+     * rules, then updates the snap sources list used by the UI
```
```suggestion
     * Creates [SnapRules] based on the given asset type, synchronizes the snap sources using these
     * rules, then updates the snap sources list used by the UI.
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+
+            // Enable snapping for the graphics overlay as this will not be affected by the given
+            // SnapSourceEnablingBehavior.setFromRules
+            sourceSettings.first { sss -> sss.source == graphicsOverlay }.isEnabled = true
```
```suggestion
            sourceSettings.first { it.source == graphicsOverlay }.isEnabled = true
```

it might be clearer in this case since sss doesn't mean too much for those who don't know the API

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+
+
+    /**
+     * Updates the enabled value of the SnapSourceSettings object at the given index and rebuilds
```
```suggestion
     * Updates the enabled value of the [SnapSourceSettings] object at the given index and rebuilds
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+        updateSnapSourceList()
+    }
+
+    /*
```
```suggestion
    /**
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+     * Updates the lists used by the UI to show SnapSourceSettings information.
+    */
+    private fun updateSnapSourceList() {
+        // Get a new list of current properties from snapSettings
+        val newSnapSourceList = currentSnapSourcePropertyList()
+
+        // Update the backing list
+        _snapSourcePropertyList.value = newSnapSourceList
```
```suggestion
        // Update the backing list with a new list of current properties from snapSettings
        _snapSourcePropertyList.value = currentSnapSourcePropertyList()
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+    }
+
+    /**
+     * Returns a list of SnapSourceProperty objects based on the current snap sources.
```
```suggestion
     * Returns a list of [SnapSourceProperty] objects based on the current snap sources.
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+     */
+    private fun resetSelections() {
+        // Clear the existing selection and show the selected feature                                      
+        (selectedFeature?.featureTable?.layer as? FeatureLayer)?.clearSelection()                          
+        (selectedFeature?.featureTable?.layer as? FeatureLayer)?.setFeatureVisible(selectedFeature!!, true)
```
```suggestion
        (selectedFeature?.featureTable?.layer as? FeatureLayer)?.let { 
            it.clearSelection()
            it.setFeatureVisible(selectedFeature!!, true)
        }
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+            )
+        }
+        val distributionPipeName = application.getString(R.string.distribution_pipe_name)
+        val servicePipeLayerName = application.getString(R.string.service_pipe_layer_name)
+        pipeLayer.subtypeSublayers.forEach { sublayer ->
```
```suggestion
        val servicePipeLayerName = application.getString(R.string.service_pipe_layer_name)
        // Assign the correct sublayer and renderer to each subtype sublayer
        pipeLayer.subtypeSublayers.forEach { sublayer ->
```

Or another comment that makes sense here

shellygill replied:
Used comment text from reference .NET implementation.

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,512 @@
+            )
+        }
+        val excessFlowValveName = application.getString(R.string.excess_flow_valve_name)
+        val controllableTeeName = application.getString(R.string.controllable_tee_name)
+        deviceLayer.subtypeSublayers.filter { sublayer ->
```
```suggestion
        val controllableTeeName = application.getString(R.string.controllable_tee_name)
        // Hide all sublayers that aren't excess flow valves or controllable tees
        deviceLayer.subtypeSublayers.filter { sublayer ->
```

TADraeseke commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/screens/SnapSourcesPanel.kt:
```diff
@@ -0,0 +1,128 @@
+            }
+        }
+
+        for (index in snapSourcePropertyList.indices) {
+            val snapSourceProperty = snapSourcePropertyList[index]
```
```suggestion
        snapSourcePropertyList.forEachIndexed { index, snapSourceProperty ->
```

TADraeseke reviewed (changes requested):
Looking great! @shubham7109 has confirmed with Sarat that we can (and I think we should) lock rotation on this sample. The UI won't really work in landscape.

Just some code and comment suggestions. Awesome sample!

shellygill commented on samples/snap-geometry-edits-with-utility-network-rules/README.md:
```diff
@@ -0,0 +1,60 @@
+1. Create a map with `LoadSettings.FeatureTilingMode` set to `EnabledWithFullResolutionWhenSupported`.
+2. Create a `Geodatabase` using the mobile geodatabase file location.
+3. Display `Geodatabase.featureTables` on the map using subtype feature layers.
+4. Create a `GeometryEditor`, set `GeometryEditor.tool` to a `ReticleVertexTool`, and connect the editor to the map view.
```
Given you've not suggested the same for creating the `GeometryEditor` or `Geodatabase` objects, I wasn't 100% clear on the thinking behind this suggestion. I've changed the text to be clear you're creating a new object - let me know if that wasnt the aim.

TADraeseke replied:
I tend to just emulate the exact syntax of the call in code so users can more easily track between the two. But it's a bit harder in Kotlin because of scoping functions. So maybe it should be:

4. Create a `GeometryEditor()` and set it's `tool` to a `ReticleVertexTool()`, and connect the editor to the map view.

Not sure. Mainly just wanted to get Async and other .NET specific function names out

shellygill reviewed (commented):
Thanks @TADraeseke I've added the suggestions/made suggested changes.


> @shubham7109 has confirmed with Sarat that we can (and I think we should) lock rotation on this sample. The UI won't really work in landscape.

Can you point me to the approach we use for this in compose samples? `android:screenOrientation` is discouraged API, and I don't see any other existing uses in this repo of this.

TADraeseke reviewed (approved):
LGTM! Addressed your comment in line. Ready for second review @shubham7109!

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+    // Hold references to the subtype sublayers for the distribution and service pipe layers
+    private var distributionPipeLayer: SubtypeSublayer? = null
+    private var servicePipeLayer: SubtypeSublayer? = null
+    private var pipelineLayerName = application.getString(R.string.pipeline_layer_name)
+    private var distributionPipeName = application.getString(R.string.distribution_pipe_name)
+    private var servicePipeLayerName = application.getString(R.string.service_pipe_layer_name)
```
Should use `val`s here:
```suggestion
    private val pipelineLayerName = application.getString(R.string.pipeline_layer_name)
    private val distributionPipeName = application.getString(R.string.distribution_pipe_name)
    private val servicePipeLayerName = application.getString(R.string.service_pipe_layer_name)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+    private var selectedFeature: ArcGISFeature? = null
+
+    private val _assetGroupNameState = MutableStateFlow("<Nothing selected>")
+    var assetGroupNameState: StateFlow<String> = _assetGroupNameState
```
This should be `val`, also I prefer to use `asStateFlow` instead of inline cast here:
```suggestion
    val assetGroupNameState = _assetGroupNameState.asStateFlow()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+    var assetGroupNameState: StateFlow<String> = _assetGroupNameState
+
+    private val _assetTypeNameState = MutableStateFlow("<Nothing selected>")
+    val assetTypeNameState: StateFlow<String> = _assetTypeNameState
```
Same as above:
```suggestion
    val assetTypeNameState = _assetTypeNameState.asStateFlow()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+        val snapSourceSettings: SnapSourceSettings
+    )
+    private val _snapSourcePropertyList = MutableStateFlow(listOf<SnapSourceProperty>())
+    internal val snapSourcePropertyList: StateFlow<List<SnapSourceProperty>> =
+        _snapSourcePropertyList
```
```suggestion
    val snapSourcePropertyList = _snapSourcePropertyList.asStateFlow()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+
+    // Create boolean flags to track the state of UI components
+    private val _isEditButtonEnabled = MutableStateFlow(false)
+    internal val isEditButtonEnabled = _isEditButtonEnabled.asStateFlow()
```
Not needed 
```suggestion
    val isEditButtonEnabled = _isEditButtonEnabled.asStateFlow()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+    init {
+        viewModelScope.launch {
+            arcGISMap.load().onFailure { error ->
+                messageDialogVM.showMessageDialog(
+                    "Failed to load map",
+                    error.message.toString()
+                )
```
Could be simplified to use the error function: 
```suggestion
                messageDialogVM.showMessageDialog(error)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                // Set the utility network on the map and load it
+                arcGISMap.utilityNetworks.add(geodatabase.utilityNetworks.first())
+                arcGISMap.utilityNetworks.first().load().onFailure {
+                    messageDialogVM.showMessageDialog(
+                        "Error loading utility network", it.message.toString()
+                    )
```
same as above: 
```suggestion
                    messageDialogVM.showMessageDialog(it)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                symbolSwatches[SnapRuleBehavior.RulesLimitSnapping] = createSwatch(rulesLimitSymbol)
+                symbolSwatches[SnapRuleBehavior.None] = createSwatch(noneSymbol)
+            }.onFailure {
+                messageDialogVM.showMessageDialog(
+                    "Error loading geodatabase", it.message.toString()
+                )
```
same as above: 
```suggestion
                messageDialogVM.showMessageDialog(it)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+        // Get the symbol for the selected feature
+        selectedFeature?.let { feature ->
+
+            // Use the symbol from the selected feature in the style of the geometry editor tool
+            val symbol =
+                (feature.featureTable as GeodatabaseFeatureTable).layerInfo?.drawingInfo?.renderer?.getSymbol(
+                    feature
+                )
```
Seems like `featureTable` will be needed in the code below, we could keep track of the value and simplify the `symbol` code a bit. Also, we need to use `as?` for all unchecked null-casts to ensure null-safety:  
```suggestion
            // Get the geodatabase feature table of the selected feature
            val featureTable = (feature.featureTable as? GeodatabaseFeatureTable) ?: return
            // Use the symbol from the selected feature in the style of the geometry editor tool
            val symbol = featureTable.layerInfo?.drawingInfo?.renderer?.getSymbol(feature)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+            }
+
+            // Hide the selected feature
+            (feature.featureTable?.layer as? FeatureLayer)?.setFeatureVisible(feature, false)
```
Using the above changes: 
```suggestion
            (featureTable.layer as? FeatureLayer)?.setFeatureVisible(feature, false)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+            feature.geometry = finalGeometry
+
+            viewModelScope.launch {
+                (feature.featureTable as GeodatabaseFeatureTable).updateFeature(feature)
+                    .onFailure { error ->
+                        messageDialogVM.showMessageDialog(
+                            "Error updating feature",
+                            error.message.toString()
+                        )
+                    }
+            }
```
Could be simplified to use (with `as?`): 
```suggestion
                (feature.featureTable as? GeodatabaseFeatureTable)?.updateFeature(feature)
                    ?.onFailure { error -> messageDialogVM.showMessageDialog(error) }
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                // contained in the sublayer results
+                val identifiedFeature = identifyResultList.firstOrNull()?.sublayerResults?.firstOrNull()?.geoElements?.firstOrNull()
+                if (identifiedFeature !is ArcGISFeature) {
+                    resetSelections()
+                    return@launch
```
Could be formatted as below and effectively is the same behavior: 
```suggestion
                    return@launch resetSelections()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                // In this sample we only allow selection of point features. If the identified
+                // feature is null or the feature is not a point feature then reset and return.
+                if (identifiedFeature.featureTable?.geometryType != GeometryType.Point) {
+                    resetSelections()
+                    return@launch
```
Same as above: 
```suggestion
                    return@launch resetSelections()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                // Update the selected feature and select it on the layer
+                selectedFeature = identifiedFeature
+                selectedFeature?.let { feature ->
+                    (feature.featureTable?.layer as FeatureLayer).selectFeature(feature)
```
Use `as?` to keep the sample null safe: 
```suggestion
                    (feature.featureTable?.layer as? FeatureLayer)?.selectFeature(feature)
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                        _assetTypeNameState.value = element.assetType.name
+                        _isEditButtonEnabled.value = true
+                        setSnapSettings(element.assetType)
+                    } ?: run {
+                        messageDialogVM.showMessageDialog("Error creating UtilityElement")
+                        return@launch
+                    }
```
Could be written as below, but feel free to take it as only a suggestion: 
```suggestion
                    } ?: return@launch messageDialogVM.showMessageDialog("Error creating UtilityElement")
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+     */
+    private suspend fun setSnapSettings(assetType: UtilityAssetType) {
+        // Get the snap rules associated with the asset type
+        val snapRules = SnapRules.create(arcGISMap.utilityNetworks.first(), assetType).getOrThrow()
```
Avoid throwing unhandled exceptions in the sample: 
```suggestion
        val snapRules = SnapRules.create(arcGISMap.utilityNetworks.first(), assetType).getOrElse {
            return messageDialogVM.showMessageDialog(it)
        }
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                ) {
+                    // If a feature is already selected and the tapped feature is not the selected
+                    // feature then clear the previous selection
+                    (selectedFeature?.featureTable?.layer as FeatureLayer).clearSelection()
```
```suggestion
                    (selectedFeature?.featureTable?.layer as? FeatureLayer)?.clearSelection()
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+                    }
+                    is SubtypeFeatureLayer -> {
+                        sourceSettings.childSourceSettings.forEach { childSourceSettings ->
+(childSourceSettings.source as? SubtypeSublayer)?.let { childSource ->
```
Formatting looks wrong here: 
```suggestion
                            (childSourceSettings.source as? SubtypeSublayer)?.let { childSource ->
```

shubham7109 commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt:
```diff
@@ -0,0 +1,514 @@
+     */
+    private fun resetSelections() {
+        // Clear the existing selection and show the selected feature                                      
+        (selectedFeature?.featureTable?.layer as? FeatureLayer)?.let { 
+            it.clearSelection()
+            it.setFeatureVisible(selectedFeature!!, true)
+        }
```
Avoid forcing nullability with `!!`: 
```suggestion
        selectedFeature?.let { feature ->
            (feature.featureTable?.layer as? FeatureLayer)?.let {
                it.clearSelection()
                it.setFeatureVisible(feature = feature, visible = true)
            }
        }
```

shubham7109 reviewed (changes requested):
@shellygill Added some view-model feedback mostly regarding nullability and some code cleanup. The same is looking great!

shellygill commented:
Thanks @shubham7109 - I have applied suggestions and re-running vTest now.

shubham7109 reviewed (approved):
Looks great, thanks for adding in the changes @shellygill!

Just one comment to lock the sample orientation, and lgtm 👍

shellygill commented on samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/MainActivity.kt:
```diff
@@ -0,0 +1,53 @@
+
+class MainActivity : ComponentActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        // authentication with an API key or named user is
+        // required to access basemaps and other location services
+        ArcGISEnvironment.apiKey = ApiKey.create(BuildConfig.ACCESS_TOKEN)
```
Ah, that's the right approach, thanks @shubham7109 !

shellygill commented:
Tested the orientation change locally, and the latest vTest is passing at https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/96/

shubham7109 commented:
Thanks @shellygill, should be good to (squash and) merge to `v.next`. This sample will be added as part of the 200.7 samples.
